Source/Core/Core/HW/GCMemcard/GCMemcard.cpp

```diff
@@ -27,196 +27,256 @@ static void ByteSwap(u8* valueA, u8* valueB)
-    case MemCard1019Mb:
-    case MemCard2043Mb:
-      break;
-    default:
-      PanicAlertT("%s failed to load as a memory card.\nCard size is invalid (0x%x bytes)",
-                  filename.c_str(), (unsigned)size);
-      return;
-    }
-  }
+bool GCMemcardErrorCode::HasCriticalErrors() const
+{
+  return Test(GCMemcardValidityIssues::FAILED_TO_OPEN) || Test(GCMemcardValidityIssues::IO_ERROR) ||
+         Test(GCMemcardValidityIssues::INVALID_CARD_SIZE) ||
+         Test(GCMemcardValidityIssues::INVALID_CHECKSUM) ||
+         Test(GCMemcardValidityIssues::MISMATCHED_CARD_SIZE) ||
+         Test(GCMemcardValidityIssues::FREE_BLOCK_MISMATCH) ||
+         Test(GCMemcardValidityIssues::DIR_BAT_INCONSISTENT);
+}
 
-  mcdFile.Seek(0, SEEK_SET);
-  if (!mcdFile.ReadBytes(&m_header_block, BLOCK_SIZE))
-  {
-    PanicAlertT("Failed to read header correctly\n(0x0000-0x1FFF)");
-    return;
-  }
-  if (m_size_mb != m_header_block.m_size_mb)
-  {
-    PanicAlertT("Memory card file size does not match the header size");
-    return;
-  }
+bool GCMemcardErrorCode::Test(GCMemcardValidityIssues code) const
+{
+  return m_errors.test(static_cast<size_t>(code));
+}
 
-  if (!mcdFile.ReadBytes(&m_directory_blocks[0], BLOCK_SIZE))
-  {
-    PanicAlertT("Failed to read 1st directory block correctly\n(0x2000-0x3FFF)");
-    return;
-  }
+void GCMemcardErrorCode::Set(GCMemcardValidityIssues code)
+{
+  m_errors.set(static_cast<size_t>(code));
+}
 
-  if (!mcdFile.ReadBytes(&m_directory_blocks[1], BLOCK_SIZE))
-  {
-    PanicAlertT("Failed to read 2nd directory block correctly\n(0x4000-0x5FFF)");
-    return;
-  }
+GCMemcardErrorCode& GCMemcardErrorCode::operator|=(const GCMemcardErrorCode& other)
+{
+  this->m_errors |= other.m_errors;
+  return *this;
+}
 
-  if (!mcdFile.ReadBytes(&m_bat_blocks[0], BLOCK_SIZE))
+GCMemcard::GCMemcard()
+    : m_valid(false), m_size_blocks(0), m_size_mb(0), m_active_directory(0), m_active_bat(0)
+{
+}
+
+std::optional<GCMemcard> GCMemcard::Create(std::string filename, u16 size_mbits, bool shift_jis)
+{
+  GCMemcard card;
+  card.m_filename = std::move(filename);
+
+  // TODO: Format() not only formats the card but also writes it to disk at m_filename.
+  // Those tasks should probably be separated.
+  if (!card.Format(shift_jis, size_mbits))
+    return std::nullopt;
+
+  return std::move(card);
+}
+
+std::pair<GCMemcardErrorCode, std::optional<GCMemcard>> GCMemcard::Open(std::string filename)
+{
+  GCMemcardErrorCode error_code;
+  File::IOFile file(filename, "rb");
+  if (!file.IsOpen())
   {
-    PanicAlertT("Failed to read 1st block allocation table block correctly\n(0x6000-0x7FFF)");
-    return;
+    error_code.Set(GCMemcardValidityIssues::FAILED_TO_OPEN);
+    return std::make_pair(error_code, std::nullopt);
   }
 
-  if (!mcdFile.ReadBytes(&m_bat_blocks[1], BLOCK_SIZE))
+  // check if the filesize is a valid memory card size
+  u16 card_size_mbits;
+  const u64 filesize = file.GetSize();
+  switch (filesize)
```

Comment:
This feels verbose and repeative...could we instead do `BytesToMegabits(filesize)` and just compare it to the list of known-good memory card sizes instead?

Reply:
Like this?